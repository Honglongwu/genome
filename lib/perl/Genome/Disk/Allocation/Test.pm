package Genome::Disk::Allocation::Test;
@ISA = qw(Exporter);
@EXPORT_OK = qw(create_group create_tmpfs_volume create_barrier spawn_child waitpids);

use strict;
use warnings;

use Carp qw(croak);
use Test::More;
use Time::HiRes qw(usleep);
use Sys::Hostname qw(hostname);
use above 'Genome';

require Genome::DataSource::LocalDataSource;

BEGIN {
    if ( $ENV{UR_DBI_NO_COMMIT} ) {
        plan skip_all => 'This test can not be run with UR_DBI_NO_COMMIT enabled!';
    }
};

sub import {
    # Don't change @_!
    my $class = $_[0];

    protect_real_data_sources();

    Genome::DataSource::LocalDataSource->import(qw(
        Genome::Disk::Allocation
        Genome::Disk::Assignment
        Genome::Disk::Group
        Genome::Disk::Volume
    ));

    # ensure test environment is loaded for _execute_system_command
    my $option = '-MGenome::Disk::Allocation::Test';
    unless (grep { $_ eq $option } @Genome::Disk::Allocation::_execute_system_command_perl5opt) {
        push @Genome::Disk::Allocation::_execute_system_command_perl5opt, $option;
    }

    $class->export_to_level(1, @_);
}


sub class_to_filename {
    my $class = shift;
    my $filename = $class;
    $filename =~ s/::/\//g;
    $filename .= '.pm';
    return $filename;
}

sub protect_real_data_sources {
    my $lds = Genome::DataSource::LocalDataSource->get();
    my $lds_path = $lds->__meta__->module_path;

    my %ds_module_filename = (
        'Genome::DataSource::Main' => 'Genome/DataSource/Main.pm',
        'Genome::DataSource::GMSchema' => 'Genome/DataSource/GMSchema.pm',
    );
    for my $ds_module (keys %ds_module_filename) {
        my $ds_module_filename = $ds_module_filename{$ds_module};

        if ($INC{$ds_module_filename}) {
            if ($INC{$ds_module_filename} eq $lds_path) {
                warn "$ds_module already protected!";
            } else {
                die "$ds_module already loaded; cannot protect it!";
            }
        }

        $INC{$ds_module_filename} = $lds_path;
        UR::Object::Type->define(
            class_name => $ds_module,
            is => $lds->class,
        );
    }
}

sub create_tmpfs {
    our %tmpfs_mnt_pts;
    my %arg = @_;
    my $size = delete $arg{size} || die;
    my $mount_path = delete $arg{path};

    unless ($mount_path) {
        $mount_path = File::Temp::tempdir('allocation_test_mount_XXXX', TMPDIR => 1);
    }

    my (undef, $fs_path) = File::Temp::tempfile('allocation_test_fs_XXXX', TMPDIR => 1, OPEN => 1);
    system("dd if=/dev/zero of=${fs_path} bs=1k count=${size}") && die "dd failed: $!";
    system("/sbin/mkfs.ext2 -F ${fs_path}") && die "mkfs failed: $!";

    my $hostname = hostname();
    system("fuseext2 -o rw+ ${fs_path} ${mount_path}") && die sprintf('%s: fuseext2 failed: %s', $hostname, $!);

    push @{$tmpfs_mnt_pts{$$}}, [$fs_path, $mount_path];

    return $mount_path;
}

sub create_tmpfs_volume {
    my %arg = @_;
    my $total_kb = delete $arg{total_kb} || die;
    my $group = delete $arg{group} || Genome::Disk::Group->get(disk_group_name => 'info_apipe');
    unless ($group) {
        $group = create_group('info_apipe');
    }

    my $mount_path = create_tmpfs(size => "${total_kb}");

    my $volume = Genome::Disk::Volume->create(
        disk_status    => 'active',
        hostname       => 'localhost',
        physical_path  => $mount_path,
        mount_path     => $mount_path,
        total_kb       => $total_kb,
        can_allocate   => 1,
    );

    my $assignment = Genome::Disk::Assignment->create(
        dg_id => $group->id,
        dv_id => $volume->id,
    );

    UR::Context->commit;

    return $volume;
}

sub create_group {
    my $disk_group_name = shift;
    my $group = Genome::Disk::Group->create(
        disk_group_name => $disk_group_name,
        permissions => '755',
        sticky => '1',
        subdirectory => $disk_group_name,
        unix_uid => 0,
        unix_gid => 0,
    );
    return $group;
}

sub create_barrier {
    my ($barrier_fh, $barrier) = File::Temp::tempfile(
        'genome_disk_allocation_barrier_XXXXXX',
        TMPDIR => 1,
        UNLINK => 0,
    );
    $barrier_fh->close();
    return $barrier;
}

sub spawn_child {
    my %args = @_;
    my $closure = delete $args{closure} || croak 'required argument missing: closure';
    my $usleep  = delete $args{usleep} || 10_000;
    my $barrier = delete $args{barrier};

    if (keys %args) {
        croak 'unexpected argument(s):', join(', ', keys %args);
    }

    my $pid = UR::Context::Process->fork();
    if (not defined $pid) {
        die 'Failed to fork()!';
    } elsif ($pid) {
        # Parent
        return $pid;
    } else {
        # Child
        while ($barrier && -e $barrier) {
            usleep($usleep);
        }
        $closure->();
        exit(0);
    }
}

sub waitpids {
    while (my $pid = shift @_) {
        my $wp_code = waitpid($pid, 0);
        if (-1 == $wp_code) {
            warn "Child pid already reaped.";
        } elsif (0 == $wp_code) {
            warn $?;
        }
    }
}

END {
    our %tmpfs_mnt_pts;
    for my $ref (@{$tmpfs_mnt_pts{$$}}) {
        my ($fs_path, $mount_path) = @$ref;

        system("fusermount -u ${mount_path}") && warn "fusermount failed";

        if (-d $mount_path) {
            system("find $mount_path -type d -delete") && warn "find -delete failed";
        }

        if (-f $fs_path) {
            system("rm ${fs_path}") && warn "rm failed";
        }
    }
};

1;
