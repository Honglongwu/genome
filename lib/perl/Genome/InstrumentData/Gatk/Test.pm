package Genome::InstrumentData::Gatk::Test;

use strict;
use warnings;

use Genome;

require Genome::Utility::Test;

class Genome::InstrumentData::Gatk::Test {
    is => 'UR::Singleton',
    has_constant => [
        test_data_dir => {
            calculate => q| return $ENV{GENOME_TEST_INPUTS}.'/Genome-InstrumentData-Gatk'; |,
        },
        reference_build => {
            calculate_from => [qw/ test_data_dir /],
            calculate => q| 
                Genome::Model::Build::ImportedReferenceSequence->__define__(
                    name => 'Test Ref Build v1',
                    data_directory => $test_data_dir.'/reference',
                ); 
            |,
        },
        bam_source => {
            calculate_from => [qw/ test_data_dir /],
            calculate => q| 
                Genome::InstrumentData::AlignmentResult::Merged->__define__(
                    id => 9999,
                    output_dir => $test_data_dir.'/inputs',
                ); 
            |,
        },
        known_indel => {
            calculate_from => [qw/ indel_result /],
            calculate => q|
                Genome::Model::Build::ImportedVariationList->__define__(
                    id => 9998,
                    indel_result => $indel_result,
                );
            |,
        },
        indel_result => {
            calculate_from => [qw/ test_data_dir /],
            calculate => q|
                Genome::Model::Tools::DetectVariants2::Result::Manual->__define__(
                    id => 9997,
                    output_dir => $test_data_dir.'/Mills_and_1000G_gold_standard/',
                );
            |,
        },
    ],
};

1;

