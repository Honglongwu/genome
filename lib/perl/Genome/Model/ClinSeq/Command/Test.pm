package Genome::Model::ClinSeq::Command::Test;

#Written by Malachi Griffith

use strict;
use warnings;
use Genome;
use Data::Dumper;
use Term::ANSIColor qw(:constants);
use Genome::Model::ClinSeq::Util qw(:all);

class Genome::Model::ClinSeq::Command::Test {
    is => 'Command::V2',
    has_input => [
        builds => { 
              is => 'Genome::Model::Build::SomaticVariation',
              is_many => 1,
              shell_args_position => 1,
              require_user_verify => 0,
              doc => 'somatic variation build(s) to summarize SVs from',
        },
        outdir => { 
              is => 'FilesystemPath',
              doc => 'Directory where output files will be written', 
        },
        _messages => {
            is => 'HASH',
            doc => 'hash of messages',
            default => {},
        },

    ],
    doc => 'summarize the SVs of somatic variation build',
};

sub help_synopsis {
    return <<EOS

genome model clin-seq test --outdir=/tmp/create_mutation_diagram/  'id in [129973671,129708625]'

EOS
}

sub help_detail {
    return <<EOS
Test code ideas 

EOS
}

sub execute {
  my $self = shift;

  #$self->append_message("1", "Test message 1");
  #$self->calculate_something();
  #$self->append_message("3", "Test message 3");

  #my %messages = %{$self->_messages};
  #print Dumper %messages;


  my @mutation_diagram_builds;
  push(@mutation_diagram_builds, Genome::Model::Build->get(129973671));
  push(@mutation_diagram_builds, Genome::Model::Build->get(129708625));

  my $mutation_diagram_cmd = Genome::Model::ClinSeq::Command::CreateMutationDiagrams->create(builds=>\@mutation_diagram_builds, outdir=>$self->outdir, collapse_variants=>1, max_snvs_per_file=>250, max_indels_per_file=>250);
  my $r = $mutation_diagram_cmd->execute();


  return 1;
}


sub append_message {
    my $self = shift;
    my $key = shift || die;
    my $value = shift || die;

    my $messages = $self->_messages;
    $messages->{$key} = $value;
    $self->_messages($messages);

    return 1;
}


sub calculate_something {
  my $self = shift;

  my $x = 2+2;
  my $message = "Test message 2.  The answer is $x";
  $self->append_message("2", $message);

  return 1;
}


1;


