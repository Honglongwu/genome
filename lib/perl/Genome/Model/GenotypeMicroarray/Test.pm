package Genome::Model::GenotypeMicroarray::Test;

use strict;
use warnings;


sub testdir {
    return $ENV{GENOME_TEST_INPUTS} . '/GenotypeMicroarray/v2';
}

my %cache;
sub reference_sequence_build {
    return $cache{reference_sequence_build} if $cache{reference_sequence_build};

    $cache{reference_sequence_build} = Genome::Model::Build::ReferenceSequence->__define__(
        name => '__TEST_REF__',
    );
    die 'Failed to define reference sequence build' if not $cache{reference_sequence_build};

    my %pos_seq = (    # Alleles
        752566 => 'A', # AG 
        752721 => 'G', # AG
        798959 => 'C', # AG SNV
        800007 => 'T', # TC
        838555 => 'C', # CC
        846808 => 'C', # CC
        854250 => 'A', # AA
        861808 => 'G', # GG DUP
        873558 => 'G', # TT SNV
        882033 => 'G', # GG
    );

    no warnings;
    *Genome::Model::Build::ReferenceSequence::sequence = sub{ return $pos_seq{$_[2]}; };
    *Genome::Model::Build::ReferenceSequence::chromosome_array_ref = sub{ return [qw/ 1 /]; };
    use warnings;

    return $cache{reference_sequence_build};
}

sub variation_list_build { # FIXME upgrade to VCF
    return $cache{variation_list_build} if $cache{variation_list_build};

    my $dbsnp_file = testdir().'/dbsnp/dbsnp.132';
    my $fl = Genome::Model::Tools::DetectVariants2::Result::Manual->__define__(
        description => '__TEST__DBSNP132__',
        username => 'apipe-tester',
        file_content_hash => 'c746fb7b7a88712d27cf71f8262dd6e8',
        output_dir => testdir().'/dbsnp',
    );
    $fl->lookup_hash($fl->calculate_lookup_hash());

    my $variation_list_model = Genome::Model::ImportedVariationList->__define__(
        name => '__TEST_DBSNP__',
        subject => Genome::Taxon->__define__(name => '__TEST_HUMAN__'),
    );

    $cache{variation_list_build} = Genome::Model::Build::ImportedVariationList->__define__(
        model => $variation_list_model,
        reference => reference_sequence_build(),
        snv_result => $fl,
        version => 'TEST',
    );
    die 'Failed to define variation list build' if not $cache{variation_list_build};

    return $cache{variation_list_build};
}

sub sample {
    return $cache{sample} if $cache{sample};
    instrument_data();
    return $cache{sample};
}

sub instrument_data {
    return $cache{instrument_data} if $cache{instrument_data};

    $cache{sample} = Genome::Sample->__define__(
        id => 2879594813,
        name => '__TEST_SAMPLE__',
    );
    die 'Failed to define sample' if not $cache{sample};

    my $library = Genome::Library->__define__(
        name => $cache{sample}->name.'-microarraylib',
        sample => $cache{sample},
    );

    $cache{instrument_data} = Genome::InstrumentData::Imported->__define__(
        library => $library,
        import_format => 'genotype file',
        sequencing_platform => 'infinium',
    );
    die 'Failed to define instrument data' if not $cache{instrument_data};
    $cache{instrument_data}->add_attribute(attribute_label => 'genotype_file', attribute_value => testdir().'/instdata/snpreport/genotypes.tsv');
    die 'Failed to set genotype file on instdata!' if not -s $cache{instrument_data}->genotype_file;

    $cache{sample}->default_genotype_data_id($cache{instrument_data}->id);
    die 'Failed to set default genotype data id on sample!' if $cache{sample}->default_genotype_data_id ne $cache{instrument_data}->id;

    return $cache{instrument_data};
}

sub model {
    return $cache{model} if $cache{model};

    my $instrument_data = instrument_data();
    my $sample = $instrument_data->sample;
    my $pp = Genome::ProcessingProfile::GenotypeMicroarray->__define__(name => '__TEST_PP__');
    $cache{model} = Genome::Model::GenotypeMicroarray->create(
        name => 'Test Genotype Microarray pp',
        processing_profile => $pp,
        subject => $sample,
        subject_id => $sample->id,
        subject_class_name => $sample->class,
        reference_sequence_build => reference_sequence_build(),
        dbsnp_build => variation_list_build(),
    );
    $cache{model}->add_instrument_data($instrument_data);

    return $cache{model};
}

sub example_build {
    return $cache{example_build} if $cache{example_build};

    my $model = model();
    $cache{example_build} = Genome::Model::Build::GenotypeMicroarray->__define__(
        model => $model,
        reference_sequence_build => $model->reference_sequence_build,
        dbsnp_build => $model->dbsnp_build,
        data_directory => testdir().'/build',
    );

    return $cache{example_build};
}

sub example_legacy_build {
    return $cache{example_legacy_build} if $cache{example_legacy_build};

    my $model = model();
    $cache{example_legacy_build} = Genome::Model::Build::GenotypeMicroarray->__define__(
        model => $model,
        reference_sequence_build => $model->reference_sequence_build,
        dbsnp_build => $model->dbsnp_build,
        data_directory => testdir().'/build-legacy',
    );
    $cache{example_legacy_build}->add_instrument_data( $model->instrument_data );

    return $cache{example_legacy_build};
}

sub build {

    my $model = model();
    my $tempdir = File::Temp::tempdir(CLEANUP => 1);
    my $build = Genome::Model::Build::GenotypeMicroarray->create(
        model => $model,
        reference => $model->reference,
        dbsnp_build => $model->dbsnp_build,
        data_directory => $tempdir,
    );
    ok($build, 'create genotype microarray build');
    $build->add_instrument_data( $model->instrument_data );

    return $build;
}

sub expected_genotypes {
    return $cache{expected_genotypes} if $cache{expected_genotypes};
    $cache{expected_genotypes} = [
        {
            'id' => 'rs3094315',
            'chromosome' => '1',
            'position' => '752566',
            'allele1' => 'A',
            'allele2' => 'G',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.8931',
            'log_r_ratio' => '-0.3639',
        },
        {
            'id' => 'rs3131972',
            'chromosome' => '1',
            'position' => '752721',
            'allele1' => 'A',
            'allele2' => 'G',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.9256',
            'log_r_ratio' => '-0.0539',
        },
        {
            'id' => 'rs11240777',
            'chromosome' => '1',
            'position' => '798959',
            'allele1' => 'A',
            'allele2' => 'G',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.8729',
            'log_r_ratio' => '-0.0192',
        },
        {
            'id' => 'rs6681049',
            'chromosome' => '1',
            'position' => '800007',
            'allele1' => 'T',
            'allele2' => 'C',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.7156',
            'log_r_ratio' => '0.2960',
        },
        {
            'id' => 'rs4970383',
            'chromosome' => '1',
            'position' => '838555',
            'allele1' => 'C',
            'allele2' => 'C',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.8749',
            'log_r_ratio' => '0.4694',
        },
        {
            'id' => 'rs4475691',
            'chromosome' => '1',
            'position' => '846808',
            'allele1' => 'C',
            'allele2' => 'C',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.8480',
            'log_r_ratio' => '-0.0174',
        },
        {
            'id' => 'rs7537756',
            'chromosome' => '1',
            'position' => '854250',
            'allele1' => 'A',
            'allele2' => 'A',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.8670',
            'log_r_ratio' => '0.0389',
            'sample_id' => '2879594813',
        },
        {
            'id' => 'rs1110052',
            'chromosome' => '1',
            'position' => '873558',
            'allele1' => 'T',
            'allele2' => 'T',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.7787',
            'log_r_ratio' => '0.1487',
        },
        {
            'id' => 'rs2272756',
            'chromosome' => '1',
            'position' => '882033',
            'allele1' => 'G',
            'allele2' => 'G',
            'cnv_confidence' => 'NA',
            'cnv_value' => '2.0',
            'gc_score' => '0.8677',
            'log_r_ratio' => '-0.0801',
        },
    ];

    my $refseq = reference_sequence_build();
    my $sam = sample();
    for my $genotype ( @{$cache{expected_genotypes}} ) {
        $genotype->{sample_id} = $sam->id;
        #$genotype->{sample_name} = $sam->name;
        $genotype->{identifiers} = [$genotype->{id}];
        $genotype->{chrom} = $genotype->{chromosome};
        $genotype->{reference} = $refseq->sequence($genotype->{chromosome}, $genotype->{position}, $genotype->{position});
        $genotype->{reference_allele} = $genotype->{reference};
        $genotype->{alleles} = $genotype->{allele1}.$genotype->{allele2};
        $genotype->{alternate_alleles} = Genome::Model::GenotypeMicroarray::GenotypeFile::ReadTsvAndAnnotate->_alts_for_genotype($genotype);
        $genotype->{quality} = '.';
        $genotype->{_filter} = [];
        $genotype->{_format} = [Genome::Model::GenotypeMicroarray::GenotypeFile::ReadTsvAndAnnotate->_format_for_genotype($genotype)];
        $genotype->{info_fields} = Genome::Model::GenotypeMicroarray::GenotypeFile::ReadTsvAndAnnotate->_info_hash_for_genotype($genotype);
        $genotype->{_sample_data} = [ [] ], #FIXME
    }

    return $cache{expected_genotypes};
}

1;

