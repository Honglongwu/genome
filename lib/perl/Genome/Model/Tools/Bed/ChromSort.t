#!/usr/bin/env genome-perl

use strict;
use warnings;

use Test::More tests => 4;
use above 'Genome';
use_ok('Genome::Model::Tools::Bed::ChromSort');

my $tmpdir = File::Temp::tempdir('Bed-ChromSort-XXXXX', DIR => "$ENV{GENOME_TEST_TEMP}/", CLEANUP => 1);
my $output_file = join('/', $tmpdir, 'output');
my $input_file = __FILE__ . '.input';
my $expected_file = __FILE__ . '.expected';

my $cmd = Genome::Model::Tools::Bed::ChromSort->create(
    input => $input_file,
    output => $output_file
    );

ok($cmd, 'Created command');
ok($cmd->execute(), 'Executed command');

my $diff = Genome::Sys->diff_file_vs_file($output_file, $expected_file);
ok(!$diff, 'output matched expected result') or diag("diff results:\n" . $diff);
