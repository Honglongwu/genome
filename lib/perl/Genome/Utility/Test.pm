use strict;
use warnings;

package Genome::Utility::Test;
use base 'Test::Builder::Module';

use Exporter 'import';
our @EXPORT_OK = qw(compare_ok sub_test run_ok capture_ok);

use Carp qw(croak);
use File::Compare qw(compare);
use IPC::System::Simple qw(capture);
use Test::More;
use File::Spec qw();


my %ERRORS = (
    'REPLACE_ARRAY_REF' => q('replace' value should be an ARRAY ref),
);
sub ERRORS {
    my ($class, $key) = @_;
    return $ERRORS{$key};
}

sub sub_test($$) {
    my ($desc, $code) = @_;
    my $tb = __PACKAGE__->builder;
    $tb->ok($code->(), $desc);
}

sub _compare_ok_parse_args {
    # First two args are always the files.
    # Then accept one scalar argument, the name.
    # And one HASH ref, the options.
    # Then validate the option names.
    my $file_1 = shift;
    my $file_2 = shift;

    my $name = (@_ % 2 == 1) ? shift : undef;
    my %o = @_;

    if ($name && $o{name}) {
        die 'duplicate name argument not expected';
    } elsif ($name && !$o{name}) {
        $o{name} = $name;
    }

    my %vo; # validated_o
    $vo{name}    = delete $o{name};
    $vo{filters} = delete $o{filters};
    $vo{replace} = delete $o{replace};
    $vo{diag}    = delete $o{diag} // 1;
    my @k = keys %o;
    if (@k) {
        croak 'unexpected options passed to compare_ok: ' . join(', ', @k);
    }

    my $filters_ref = ref($vo{filters});
    if (defined $vo{filters} && (!$filters_ref || $filters_ref ne 'ARRAY')) {
        $vo{filters} = [$vo{filters}];
    }

    my $replace_ref = ref($vo{replace});
    if (defined $vo{replace} && (!$replace_ref || $replace_ref ne 'ARRAY')) {
        die sprintf(q(%s: %s\n), $ERRORS{REPLACE_ARRAY_REF}, $vo{replace});
    }

    return ($file_1, $file_2, %vo);
}

sub compare_ok {
    my ($file_1, $file_2, %o) = _compare_ok_parse_args(@_);

    my $tb = __PACKAGE__->builder;

    my @compare_args = (
        $file_1,
        $file_2,
        sub {
            for my $pr (@{$o{replace}}) {
                my ($pattern, $replacement) = @{$pr};
                map { $_ =~ s/$pattern/$replacement/ } @_;
            }
            for my $filter (@{$o{filters}}) {
                map { $_ =~ s/$filter//g } @_;
            }
            my $c = ($_[0] ne $_[1]);
            if ($c == 1 && $o{diag}) {
                $tb->diag("First diff:\n--- " . $file_1 . "\n+++ " . $file_2 . "\n- " . $_[0] . "+ " . $_[1]);
            }
            return $c;
        }
    );

    return $tb->ok(compare(@compare_args) == 0, $o{name});
}

sub capture_ok {
    my ($command, $test_name) = @_;

    my $tb = __PACKAGE__->builder;

    my @command = ref $command ? @$command : $command;
    $test_name //= @command > 1 ? join(' ', @command) : $command[0];

    my @output = eval { capture(@command) };
    my $error = $@;
    my $exit_zero = ($? == 0);
    $tb->ok($exit_zero, $test_name) or diag $error, @output;

    if (wantarray) {
        return ($exit_zero, @output);
    } else {
        return $exit_zero;
    }
}

sub run_ok {
    my ($command, $test_name) = @_;

    my $tb = __PACKAGE__->builder;

    my @command = ref $command ? @$command : $command;
    $test_name //= @command > 1 ? join(' ', @command) : $command[0];

    my $exit_zero = (system(@command) == 0);
    $tb->ok($exit_zero, $test_name);

    return $exit_zero;
}

sub data_dir_ok {
    my ($class, $package) = @_;
    (my $dirname = $package) =~ s/::/-/g;
    my $dirpath = File::Spec->join($ENV{GENOME_TEST_INPUTS}, $dirname);
    my $tb = __PACKAGE__->builder;
    $tb->ok(-d $dirpath, "data_dir exists: $dirpath");
    return $dirpath;
}

1;

__END__

=pod

=head1 NAME

Genome::Utility::Test

=head1 SYNOPSIS

    use Genome::Utiltiy::Test qw(compare_ok sub_test);

    sub_test('this diffs something' => sub {
        compare_ok($file_1, $file_1);
    });

=head1 METHODS

=item sub_test

Mimics Test::More's subtest since Ubuntu 10.04, which we run, does not have a
Test::More recent enough to have subtest support.

=item compare_ok

compare_ok use File::Compare with a few conveniences.

compare_ok($file_1, $file_2, name => '', diag => 0, test => 0);

=over4

=head2 OPTIONS

=item name

Specify a name for a test but requires 'name' key.

=item diag

Disable diag output when a diff is encountered. Added this in case people want to surpress any output.

=item test

Disable test usage, just return status. Added this so I could test compare_ok.
