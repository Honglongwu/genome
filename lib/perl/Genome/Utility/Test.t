use strict;
use warnings;

use Test::More;
use above 'Genome';

BEGIN {
    use_ok 'Genome::Utility::Test', qw(sub_test compare_ok);
}

sub_test('compare_ok' => sub {
    my $a_fh = File::Temp->new(TMPDIR => 1);
    my $a_fn = $a_fh->filename;
    $a_fh->print("a\n");
    $a_fh->close();

    my $b_fh = File::Temp->new(TMPDIR => 1);
    my $b_fn = $b_fh->filename;
    $b_fh->print("b\n");
    $b_fh->close();

    my $aa_fh = File::Temp->new(TMPDIR => 1);
    my $aa_fn = $aa_fh->filename;
    $aa_fh->print("a\n"); # like a, not aa!
    $aa_fh->close();

    {
        my $compare_ok = compare_ok($a_fn, $b_fn, test => 0);
        my $diff    = (system(qq(diff -u "$a_fn" "$b_fn" > /dev/null)) == 0 ? 1 : 0);
        is($diff, 0, 'diff detected diff between different files');
        is($compare_ok, $diff, 'compare_ok detected diff between different files');
    }

    {
        my $compare_ok = compare_ok($a_fn, $aa_fn, test => 0);
        my $diff    = (system(qq(diff -u "$a_fn" "$aa_fn" > /dev/null)) == 0 ? 1 : 0);
        is($diff, 1, 'diff did not detect diff between similar files');
        is($compare_ok, $diff, 'compare_ok did not detect diff between similar files');
    }
});

done_testing();
