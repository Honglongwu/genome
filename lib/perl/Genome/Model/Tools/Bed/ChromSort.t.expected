../../../../../../test-data-relocate-me/lib--perl--Genome--Model--Tools--Bed--ChromSort.t.expected